File: backend/src/services/provider.service.js
const { Provider, Hospital, Plan, Review, User } = require('../models');
const { Op } = require('sequelize');

class ProviderService {
  async getAllProviders(filters = {}) {
    try {
      const {
        name,
        city,
        state,
        rating,
        minPlans,
        maxPlans,
        page = 1,
        limit = 10
      } = filters;

      const where = { isActive: true };

      if (name) {
        where.name = { [Op.iLike]: `%${name}%` };
      }

      if (city) {
        where.city = { [Op.iLike]: `%${city}%` };
      }

      if (state) {
        where.state = { [Op.iLike]: `%${state}%` };
      }

      if (rating) {
        where.rating = { [Op.gte]: rating };
      }

      const offset = (page - 1) * limit;

      const { rows: providers, count } = await Provider.findAndCountAll({
        where,
        include: [{
          model: Plan,
          as: 'plans',
          attributes: ['id', 'name', 'price', 'coverage'],
          where: minPlans || maxPlans ? {
            price: {
              ...(minPlans && { [Op.gte]: minPlans }),
              ...(maxPlans && { [Op.lte]: maxPlans })
            }
          } : undefined
        }, {
          model: Hospital,
          as: 'hospitals',
          attributes: ['id', 'name', 'city', 'state']
        }],
        order: [
          ['rating', 'DESC'],
          ['name', 'ASC']
        ],
        limit,
        offset,
        distinct: true
      });

      return {
        providers,
        pagination: {
          total: count,
          page: parseInt(page),
          totalPages: Math.ceil(count / limit)
        }
      };
    } catch (error) {
      throw new Error(`Error fetching providers: ${error.message}`);
    }
  }

  async getProviderById(providerId) {
    try {
      const provider = await Provider.findByPk(providerId, {
        include: [
          {
            model: Plan,
            as: 'plans',
            attributes: ['id', 'name', 'price', 'coverage', 'features', 'benefits']
          },
          {
            model: Hospital,
            as: 'hospitals',
            attributes: ['id', 'name', 'address', 'city', 'state', 'rating']
          },
          {
            model: Review,
            as: 'reviews',
            include: [{
              model: User,
              as: 'user',
              attributes: ['id', 'firstName', 'lastName', 'avatar']
            }],
            limit: 5,
            order: [['createdAt', 'DESC']]
          }
        ]
      });

      if (!provider) {
        throw new Error('Provider not found');
      }

      // Get provider stats
      const stats = await this._getProviderStats(providerId);

      return {
        ...provider.toJSON(),
        stats
      };
    } catch (error) {
      throw new Error(`Error fetching provider details: ${error.message}`);
    }
  }

  async getProviderReviews(providerId, page = 1, limit = 10) {
    try {
      const offset = (page - 1) * limit;

      const { rows: reviews, count } = await Review.findAndCountAll({
        where: { providerId },
        include: [{
          model: User,
          as: 'user',
          attributes: ['id', 'firstName', 'lastName', 'avatar']
        }],
        order: [['createdAt', 'DESC']],
        limit,
        offset
      });

      return {
        reviews,
        pagination: {
          total: count,
          page: parseInt(page),
          totalPages: Math.ceil(count / limit)
        }
      };
    } catch (error) {
      throw new Error(`Error fetching provider reviews: ${error.message}`);
    }
  }

  async addProviderReview(providerId, userId, data) {
    try {
      const { rating, comment } = data;

      // Check if user has an active plan with this provider
      const hasActivePlan = await this._checkUserHasActivePlan(userId, providerId);
      if (!hasActivePlan) {
        throw new Error('Only users with active plans can review providers');
      }

      const review = await Review.create({
        providerId,
        userId,
        rating,
        comment
      });

      // Update provider average rating
      await this._updateProviderRating(providerId);

      return review;
    } catch (error) {
      throw new Error(`Error adding provider review: ${error.message}`);
    }
  }

  async getProviderPlans(providerId) {
    try {
      const plans = await Plan.findAll({
        where: {
          providerId,
          isActive: true
        },
        order: [
          ['price', 'ASC']
        ]
      });

      return plans;
    } catch (error) {
      throw new Error(`Error fetching provider plans: ${error.message}`);
    }
  }

  async getProviderHospitals(providerId, filters = {}) {
    try {
      const {
        city,
        state,
        rating,
        page = 1,
        limit = 10
      } = filters;

      const where = {
        providerId,
        isActive: true
      };

      if (city) {
        where.city = { [Op.iLike]: `%${city}%` };
      }

      if (state) {
        where.state = { [Op.iLike]: `%${state}%` };
      }

      if (rating) {
        where.rating = { [Op.gte]: rating };
      }

      const offset = (page - 1) * limit;

      const { rows: hospitals, count } = await Hospital.findAndCountAll({
        where,
        order: [
          ['rating', 'DESC'],
          ['name', 'ASC']
        ],
        limit,
        offset
      });

      return {
        hospitals,
        pagination: {
          total: count,
          page: parseInt(page),
          totalPages: Math.ceil(count / limit)
        }
      };
    } catch (error) {
      throw new Error(`Error fetching provider hospitals: ${error.message}`);
    }
  }

  // Private helper methods
  async _getProviderStats(providerId) {
    try {
      const [
        plansCount,
        hospitalsCount,
        subscribersCount,
        averageRating
      ] = await Promise.all([
        Plan.count({ where: { providerId, isActive: true } }),
        Hospital.count({ where: { providerId, isActive: true } }),
        User.count({
          include: [{
            model: Plan,
            as: 'activePlan',
            where: { providerId }
          }]
        }),
        Review.findOne({
          where: { providerId },
          attributes: [
            [sequelize.fn('AVG', sequelize.col('rating')), 'averageRating']
          ]
        })
      ]);

      return {
        plansCount,
        hospitalsCount,
        subscribersCount,
        averageRating: averageRating?.getDataValue('averageRating') || 0
      };
    } catch (error) {
      throw new Error(`Error fetching provider stats: ${error.message}`);
    }
  }

  async _checkUserHasActivePlan(userId, providerId) {
    const user = await User.findByPk(userId, {
      include: [{
        model: Plan,
        as: 'activePlan',
        where: { providerId },
        required: true
      }]
    });

    return !!user;
  }

  async _updateProviderRating(providerId) {
    const avgRating = await Review.findOne({
      where: { providerId },
      attributes: [
        [sequelize.fn('AVG', sequelize.col('rating')), 'averageRating']
      ]
    });

    await Provider.update(
      { rating: avgRating.getDataValue('averageRating') },
      { where: { id: providerId } }
    );
  }
}

module.exports = new ProviderService(); 

File: lib/services/provider_service.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import '../models/provider_model.dart';

class ProviderService extends ChangeNotifier {
  final String _baseUrl = '${ApiConfig.baseUrl}/providers';
  List<ProviderModel> _providers = [];
  ProviderModel? _selectedProvider;
  bool _isLoading = false;
  String? _error;

  List<ProviderModel> get providers => _providers;
  ProviderModel? get selectedProvider => _selectedProvider;
  bool get isLoading => _isLoading;
  String? get error => _error;

  Future<ProviderListResponse> getProviders(Map<String, dynamic> filters) async {
    try {
      final queryParams = {
        if (filters['name'] != null) 'name': filters['name'],
        if (filters['city'] != null) 'city': filters['city'],
        if (filters['state'] != null) 'state': filters['state'],
        if (filters['rating'] != null) 'rating': filters['rating'].toString(),
        if (filters['minPlans'] != null) 'minPlans': filters['minPlans'].toString(),
        if (filters['maxPlans'] != null) 'maxPlans': filters['maxPlans'].toString(),
        'page': (filters['page'] ?? 1).toString(),
        'limit': (filters['limit'] ?? 10).toString(),
      };

      final uri = Uri.parse(_baseUrl).replace(queryParameters: queryParams);
      final response = await http.get(uri);

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success']) {
          final List<ProviderModel> providers = (data['data'] as List)
              .map((json) => ProviderModel.fromJson(json))
              .toList();

          if (filters['page'] == 1) {
            _providers = providers;
          } else {
            _providers.addAll(providers);
          }

          notifyListeners();

          return ProviderListResponse(
            providers: providers,
            pagination: PaginationData(
              total: data['pagination']['total'],
              page: data['pagination']['page'],
              totalPages: data['pagination']['totalPages'],
            ),
          );
        }
      }
      throw ApiException('Failed to fetch providers');
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      rethrow;
    }
  }

  Future<ProviderModel> getProviderDetails(String providerId) async {
    try {
      _isLoading = true;
      _error = null;
      notifyListeners();

      final response = await http.get(Uri.parse('$_baseUrl/$providerId'));

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success']) {
          _selectedProvider = ProviderModel.fromJson(data['data']);
          notifyListeners();
          return _selectedProvider!;
        }
      }
      throw ApiException('Failed to fetch provider details');
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      rethrow;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<List<Map<String, dynamic>>> getProviderReviews(
    String providerId, {
    int page = 1,
    int limit = 10,
  }) async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/$providerId/reviews').replace(
          queryParameters: {
            'page': page.toString(),
            'limit': limit.toString(),
          },
        ),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success']) {
          return List<Map<String, dynamic>>.from(data['data']);
        }
      }
      throw ApiException('Failed to fetch provider reviews');
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      rethrow;
    }
  }

  Future<void> addProviderReview(
    String providerId, {
    required int rating,
    String? comment,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/$providerId/reviews'),
        body: json.encode({
          'rating': rating,
          'comment': comment,
        }),
        headers: {
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 201) {
        final data = json.decode(response.body);
        if (data['success']) {
          // Refresh provider details to update rating
          await getProviderDetails(providerId);
          return;
        }
      }
      throw ApiException('Failed to add review');
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      rethrow;
    }
  }

  Future<List<Map<String, dynamic>>> getProviderPlans(String providerId) async {
    try {
      final response = await http.get(Uri.parse('$_baseUrl/$providerId/plans'));

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success']) {
          return List<Map<String, dynamic>>.from(data['data']);
        }
      }
      throw ApiException('Failed to fetch provider plans');
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      rethrow;
    }
  }

  Future<List<Map<String, dynamic>>> getProviderHospitals(
    String providerId, {
    Map<String, dynamic>? filters,
  }) async {
    try {
      final queryParams = {
        if (filters?['city'] != null) 'city': filters!['city'],
        if (filters?['state'] != null) 'state': filters!['state'],
        if (filters?['rating'] != null) 'rating': filters!['rating'].toString(),
        'page': (filters?['page'] ?? 1).toString(),
        'limit': (filters?['limit'] ?? 10).toString(),
      };

      final response = await http.get(
        Uri.parse('$_baseUrl/$providerId/hospitals').replace(
          queryParameters: queryParams,
        ),
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        if (data['success']) {
          return List<Map<String, dynamic>>.from(data['data']);
        }
      }
      throw ApiException('Failed to fetch provider hospitals');
    } catch (e) {
      _error = e.toString();
      notifyListeners();
      rethrow;
    }
  }

  void clearError() {
    _error = null;
    notifyListeners();
  }
}

class ProviderListResponse {
  final List<ProviderModel> providers;
  final PaginationData pagination;

  ProviderListResponse({
    required this.providers,
    required this.pagination,
  });
}

class PaginationData {
  final int total;
  final int page;
  final int totalPages;

  PaginationData({
    required this.total,
    required this.page,
    required this.totalPages,
  });
} 

class ApiException implements Exception {
  final String message;
  ApiException(this.message);

  @override
  String toString() => 'ApiException: $message';
}